src/librustc/ty/maps/on_disk_cache.rs

```diff
@@ -289,6 +499,17 @@ impl<'enc, 'tcx, E> SpecializedEncoder<ty::GenericPredicates<'tcx>>
     }
 }
 
+// NodeIds are not stable across compilation sessions, so we store them in their
+// HirId representation. This allows use to map them to the current NodeId.
+impl<'enc, 'tcx, E> SpecializedEncoder<NodeId> for CacheEncoder<'enc, 'tcx, E>
+    where E: 'enc + ty_codec::TyEncoder
+{
+    fn specialized_encode(&mut self, node_id: &NodeId) -> Result<(), Self::Error> {
+        let hir_id = self.definitions.node_to_hir_id(*node_id);
```

Comment:
Is this basically copy-n-paste from the other encoders? (Not criticizing, just trying to understand.)

Reply:
Answer: this impl, anyway, is not.

Reply:
Yeah, I thought about writing a macro in `ty::codec` that would instantiate these implementations. Upfront, I wasn't sure though how many of them would end up just being copies.